### DISCUSSION TITLE
Chained constructor calls are far too restrictive

### DISCUSSION THREAD

**Discussion by masonwheeler** [dotnet/csharplang#5845] [Language Ideas] [answered]
C#'s constructor chaining syntax (the call to `: this()` or `: base()` constructors at the start of the constructor) is lifted from C++, despite the problem that that syntax exists to solve (C++'s broken, upside-down object construction model) not existing in the CLR.

The `this` object reference exists, in fully-realized, valid form, before a chained constructor call ever executes, as proven by the existence of field initializers.  They run before the chained constructor invocation, and are able to assign values to fields of the current object.  Likewise, vtables appear to be properly set up with no problems before a chained constructor is called.  (This can be verified by placing a breakpoint on a field initializer, then calling a virtual method in Immediate Window when the breakpoint is hit.)  However, the `this` *keyword* is arbitrarily not valid until after the chained constructor has run.

Under C++ this makes sense.  In C# it does not, and the restriction should be loosened to a point that does make sense.

For example, in the Boo language, the rules are significantly more permissive than C#'s.  A field initializer may refer to fields that have already been initialized (with order of operations defined as a strict top-to-bottom source code ordering) but not to fields that have not yet been initialized, and the chained constructor call is written as a simple method call inside the method body.  It must come before any method calls inside the method body, but other things (such as setting things up with local variables) can come before it.  These rules work just fine; they are restrictive enough to not cause object-initialization screwups, without forbidding things that aren't a problem anyway under the .NET object model.

The specific pain point here involves setting up a cache that lazily loads/constructs missing entries.  My use case looks something like this:

```
public class AbstractCacheHolder
{
   protected AbstractCacheHolder(Func<Value1, Value2> missingValueGenerator)
   {
      //set up the cache, using missingValueGenerator to lazily load cache entries on demand
   }
}

public class MyCacheHolder : AbstractCacheHolder
{
   public MyCacheHolder() : base(this.GenerationFunction) //arbitrarily not allowed to do this
   {}

   private Value2 GenerationFunction(Value1 key)
   {
      // logic goes here
   }
}
```

`AbstractCacheHolder` is external code; I'm not able to modify its design.  And the obvious way to implement the cache I'm working on... is disallowed for no particularly good reason.  I'd like to see that changed.

**Comment by TonyValenti** [comment]
I really like the way that VB chains constructors as opposed to C#. 

For your specific case, make the method you're calling in your constructor a static method and it will work.

**Reply by HaloFour** [reply]
The appreciable difference is that VB.NET emits field initializers after the call to the base constructor, so it's legal to reference instance members.  But VB.NET still forbids the use case stated here.

**Reply by masonwheeler** [reply]
That doesn't work, as the method in the actual use case requires access to instance data.  (If it was that simple, I wouldn't need to be asking for language fixes to enable it!)

**Comment by HaloFour** [comment, marked as answer]
See: https://github.com/dotnet/csharplang/discussions/5813

**Reply by 333fred** [reply]
Marking as the answer as this is a duplicate of the same recently-suggested change.
